lib/puppet/provider/group/windows_adsi.rb

```diff
@@ -24,7 +24,7 @@ def members_insync?(current, should)
     # since the default array_matching comparison is not commutative
 
     # dupes automatically weeded out when hashes built
-    current_users = Puppet::Util::Windows::ADSI::Group.name_sid_hash(current)
+    current_users = Hash[ current.map { |u| [ u.sid, u ] } ]
```

Comment:
This could probably be cleaned up a bit to make it easier to read:

We use current to split out the sid and name, only to just grab the keys later and ignore the name.

Similar with `specified_users`.  Assigns hash and then only uses the keys

from
``` ruby
current_users = Hash[ current.map { |u| [ u.sid, u ] } ]
specified_users = Puppet::Util::Windows::ADSI::Group.name_sid_hash(should)
  
current_sids = current_users.keys.to_a
specified_sids = specified_users.keys.to_a
```
to
``` ruby
current_sids = Array[ current.map { |u| [ u.sid ] } ]
specified_sids = Puppet::Util::Windows::ADSI::Group.name_sid_hash(should).keys.to_a
```
(or something to that effect)

Reply:
On phone now/ will take another look later, but I'm pretty sure the hashes are necessary here for the set math / dupe removal (see related tests)

Reply:
So looking at this again... 

* For `current`, I don't think it's possible for ADSI to return a list of non-unique group members, as that is not permitted in Windows AFAIK.... so we can probably simplify that part to `current_sids = current.map(&:sid)`
* For `specified_users`, there aren't any guardrails for end users preventing them from supplying duplicate names in the members list. That said, `name_sid_hash` removes any dupes already I believe... so `specified_sids` should be able to be set like you've suggested.

I'll look into adding a follow-on commit that refactors / simplifies this.

Thanks!